Add form_with to unify form_tag/form_for.

kaspth opened the pull request:
Ref https://github.com/rails/rails/issues/25197

`form_tag` and `form_for` serve very similar use cases. This
PR unifies that usage such that `form_with` can output just
the opening form tag akin to `form_tag` and can just work with
a url, for instance.

`form_with` by default doesn't attach a class or id to the form.

Ported over old tests where applicable to ensure maximum coverage,
but left some commented out because they don't yet apply (e.g.
`fields_for` later being replaced by `fields`).

Pending in this PR:
- [x] Add new `fields` DSL
- [x] Default to `remote: true`
- [x] Remove commented out tests in the `form_with_test.rb`.
- [x] Basic documentation
- [x] Changelog entry 😁

Later work up for grabs once this is in:
- Add overridable field values, e.g. `form.text_field :name, 'override'`
- Allow fields not on the model, e.g. `form.text_field :not_implemented_on_model`
- Revisit form field defaults e.g. stop outputting ids and classes on fields by default
- Revisit form options DSL methods
- Implement `form_tag` and `form_for` through `form_with` (extract the implementations to a common object and set configs)

kaspth committed:
Add form_with to unify form_tag/form_for.

`form_tag` and `form_for` serve very similar use cases. This
PR unifies that usage such that `form_with` can output just
the opening form tag akin to `form_tag` and can just work with
a url, for instance.

`form_with` by default doesn't attach class or id to the form —
removing them on fields is moved out to a default revisiting PR later.

Ported over old tests where applicable to ensure maximum coverage,
but left some commented out because they don't yet apply (e.g.
`fields_for` later being replaced by `fields`).

[ Kasper Timm Hansen & Marek Kirejczyk ]

kaspth committed:
Code climatize.

kaspth committed:
Add fields DSL method.

Strips `_for` and requires models passed as a keyword argument.

kaspth commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -1586,6 +1662,11 @@ def fields_for(record_name, record_object = nil, fields_options = {}, &block)
+      # TODO: Documentation
+      def fields(scope = nil, model: nil, **options, &block)
+        fields_for(scope || model, model, **options, &block)
+      end
```
@dhh added the fields method to do:

```ruby
form_with(model: @post) do |form|
  form.fields(:comment, model: @comment) do |fields|
    fields.text_field # ...
  end
end
```

But it reads a bit weird if the passed `model` is a collection: `fields(:comments, model: @post.comments)`.

Not sure if you intended this to be much more than a `fields_for` alias with the don't output ids and classes by default config.

kaspth committed:
Document form_with.

Graft the `form_for` docs: rewrite, revise and expand where
needed.

Also test that a `format` isn't used when an explicit URL
is passed.

kaspth commented:
Documented `form_with` and had one hell of a time doing it! Really got to know the existing API and where the new one differs, plus in what way it does.

Will add docs for `fields` tomorrow.

kaspth committed:
Enable remote by default.

Brand new world! Forms submit via XHRs by default, woah.

kaspth commented:
Figured since `form_with` already doesn't output `id` and `class` by default, we may as well rejigger the remaining default at the form level: `remote`. It's now on by default!

kaspth committed:
Readd commented out tests.

Gives us something to revise when we're redoing the
form options helpers.

Also deletes the needless tests for the unsupported namespace
option.

kaspth commented:
This should be everything that's needed to prevent the default ids:

```patch
diff --git a/actionview/lib/action_view/helpers/form_helper.rb b/actionview/lib/action_view/helpers/form_helper.rb
index ef114b3..8550ad8 100644
--- a/actionview/lib/action_view/helpers/form_helper.rb
+++ b/actionview/lib/action_view/helpers/form_helper.rb
@@ -703,7 +703,7 @@ def form_with(model: nil, scope: nil, url: nil, format: nil, html: {}, remote: t
         html_options[:remote] = remote unless html_options.key?(:remote)
 
         if block_given?
-          builder = instantiate_builder(scope, model, options)
+          builder = instantiate_builder(scope, model, options.merge(skip_default_ids: true))
           output  = capture(builder, &Proc.new)
           html_options[:multipart] ||= builder.multipart?
 
@@ -963,12 +963,7 @@ def fields_for(record_name, record_object = nil, options = {}, &block)
 
       # TODO: Documentation
       def fields(scope = nil, model: nil, **options, &block)
-        # TODO: Remove when ids and classes are no longer output by default.
-        if model
-          scope ||= model_name_from_record_or_class(model).param_key
-        end
-
-        builder = instantiate_builder(scope, model, options)
+        builder = instantiate_builder(scope, model, options.merge(skip_default_ids: true))
         capture(builder, &block)
       end
 
@@ -1537,7 +1532,7 @@ def to_model
       def initialize(object_name, object, template, options)
         @nested_child_index = {}
         @object_name, @object, @template, @options = object_name, object, template, options
-        @default_options = @options ? @options.slice(:index, :namespace) : {}
+        @default_options = @options ? @options.slice(:index, :namespace, :skip_default_ids) : {}
         if @object_name.to_s.match(/\[\]$/)
           if (object ||= @template.instance_variable_get("@#{Regexp.last_match.pre_match}")) && object.respond_to?(:to_param)
             @auto_index = object.to_param
@@ -1840,6 +1835,7 @@ def fields_for(record_name, record_object = nil, fields_options = {}, &block)
 
       # TODO: Documentation
       def fields(scope = nil, model: nil, **options, &block)
+        options[:skip_default_ids] = true
         fields_for(scope || model, model, **options, &block)
       end
 
diff --git a/actionview/lib/action_view/helpers/tags/base.rb b/actionview/lib/action_view/helpers/tags/base.rb
index cf8a6d6..b8c446c 100644
--- a/actionview/lib/action_view/helpers/tags/base.rb
+++ b/actionview/lib/action_view/helpers/tags/base.rb
@@ -13,6 +13,7 @@ def initialize(object_name, method_name, template_object, options = {})
 
           @object_name.sub!(/\[\]$/, "") || @object_name.sub!(/\[\]\]$/, "]")
           @object = retrieve_object(options.delete(:object))
+          @skip_default_ids = options.delete(:skip_default_ids)
           @options = options
           @auto_index = Regexp.last_match ? retrieve_autoindex(Regexp.last_match.pre_match) : nil
         end
@@ -81,9 +82,12 @@ def add_default_name_and_id_for_value(tag_value, options)
           def add_default_name_and_id(options)
             index = name_and_id_index(options)
             options["name"] = options.fetch("name") { tag_name(options["multiple"], index) }
-            options["id"] = options.fetch("id") { tag_id(index) }
-            if namespace = options.delete("namespace")
-              options["id"] = options["id"] ? "#{namespace}_#{options['id']}" : namespace
+
+            unless skip_default_ids?
+              options["id"] = options.fetch("id") { tag_id(index) }
+              if namespace = options.delete("namespace")
+                options["id"] = options["id"] ? "#{namespace}_#{options['id']}" : namespace
+              end
             end
           end
 
@@ -154,6 +158,10 @@ def add_options(option_tags, options, value = nil)
           def name_and_id_index(options)
             options.key?("index") ? options.delete("index") || "" : @auto_index
           end
+
+          def skip_default_ids?
+            @skip_default_ids
+          end
       end
     end
   end
```

I think there's likely more code we can skip in the tags.

kaspth committed:
[ci skip] Document the fields helpers.

Treat both the FormBuilder and FormHelper.

kaspth commented:
Allright, I think this is ready for review 😁 — I'll merge this tomorrow/tuesday/later depending on the feedback.

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #     <input type="text" name="post[title]" value="<the title of the post>">
+      #   </form>
+      #
+      # The parameters in the forms are accessible in controlleres according to
```
`controllers`

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #
+      # For ease of comparison the examples above left out the submit button,
+      # as well as the auto generated hidden fields that enable UTF-8 support
+      # and adds an authenticity token needed for Cross Site Request Forgery
```
Don't think this needs title case. It is referenced as small case in its own documentation.

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #
+      # ==== +form_with+ options
+      #
+      # * <tt>:url</tt> - The URL the form submits to. Akin to values passed to
```
I would use "similar" instead of "Akin", since its easier to understand.

kaspth replied:
Akin seems fine to me.

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #   either "get" or "post". If "patch", "put", "delete", or another verb
+      #   is used, a hidden input named <tt>_method</tt> is added to
+      #   simulate the verb over post.
+      # * <tt>:format</tt> - The format of the route post submits to.
```
post, put, patch, etc

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      # * <tt>:scope</tt> - The scope to prefix input field names with and
+      #   thereby how the submitted parameters are grouped in controllers.
+      # * <tt>:model</tt> - A model object to infer the <tt>:url</tt> and
+      #   <tt>:scope</tt> by plus fill out input field values.
```
either comma after "by", or "and" instead of "by".

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #   E.g. turn <tt>params[:post]</tt> into <tt>params[:article]</tt>.
+      # * <tt>:authenticity_token</tt> - Authenticity token to use in the form.
+      #   Override with a custom authenticity token or pass <tt>false</tt> to
+      #   skip the authenticity_token field altogether.
```
`<tt>:authenticity_token</tt>`

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #   <form action="http://www.example.com" method="post" data-behavior="autosave" name="go">
+      #     <input name="_method" type="hidden" value="patch" />
+      #     ...
+      #   </form>
```
Looks like wrong output?

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      # Some ORM systems do not use IDs on nested models so in this case you want to be able
+      # to disable the hidden id.
+      #
+      # In the following example the Post model has many Comments stored within it in a NoSQL database,
```
it, in

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -720,6 +961,235 @@ def fields_for(record_name, record_object = nil, options = {}, &block)
+      # Writers are considered a nested attributes setter if they're of the
+      # <tt>*_attributes=</tt> form, e.g. <tt>address_attributes=</tt>.
+      #
+      # Depending on the association's reader method return value a different
```
value, a

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -720,6 +961,235 @@ def fields_for(record_name, record_object = nil, options = {}, &block)
+      # <tt>*_attributes=</tt> form, e.g. <tt>address_attributes=</tt>.
+      #
+      # Depending on the association's reader method return value a different
+      # form builder is yielded. For single object returns a one-to-one builder,
```
object, returns

vipulnsward commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -1586,6 +2056,229 @@ def fields_for(record_name, record_object = nil, fields_options = {}, &block)
+      # Note +fields+ automatically generates a hidden field to store the record
+      # ID. For circumstances where this is not needed pass
+      # <tt>include_id: false</tt> to skip it.
+      def fields(scope = nil, model: nil, **options, &block)
```
I believe this needs the same fixes from above.

kaspth replied:
Yup, copied to mirror how `fields_for` documents itself (i.e. both as the helper and in the builder). I'll update this to just refer to the `fields` helper docs above.

dhh commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #
+      # By default +form_with+ attaches the <tt>data-remote</tt> attribute
+      # submitting the form via an XMLHTTPRequest in the background if an
+      # an Unobtrusive JavaScript driver, like jquery-ujs, is used. See the
```
Since we're actually dropping jquery-ujs, maybe we need to tell a different story here.

kaspth replied:
I thought we we're just dropping the jquery in jquery-ujs. Isn't the plan to keep the ujs behavior around?

dhh replied:
Yeah, UJS is still there. Just that referencing jquery-ujs seems dated when that won't be the default.

kaspth replied:
Got it 👍

dhh commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #   <tt>config.action_view.embed_authenticity_token_in_remote_forms = false</tt>.
+      #   This is helpful when fragment-caching the form. Remote forms
+      #   get the authenticity token from the <tt>meta</tt> tag, so embedding is
+      #   unnecessary unless you support browsers without JavaScript.
```
Since our new default is remote: true, I think we need to think this through a little more. Like having authenticity_token be reliant on that value. If you do remote: false, THEN we'll by default include authenticity_token, but otherwise not. And in both cases you can always supply one via the option to overwrite the default.

kaspth replied:
Yup, figured it was just flipping the default. But of course there's repercussions.

dhh replied:
True. Guess it can just ignore those embedded tokens. I was thinking that remote: true would force the form remote, but of course, if JS is off, that won't be read.

dhh commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #   unnecessary unless you support browsers without JavaScript.
+      # * <tt>:remote</tt> - Set to true to allow the Unobtrusive
+      #   JavaScript drivers to control the submit behavior, defaulting to
+      #   to an XHR submit. Disable with <tt>remote: false</tt>.
```
The default is true, so needs to flip what this is for.

kaspth replied:
Ah yes!

dhh commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #   JavaScript drivers to control the submit behavior, defaulting to
+      #   to an XHR submit. Disable with <tt>remote: false</tt>.
+      # * <tt>:enforce_utf8</tt> - If set to false, a hidden input with name
+      #   utf8 is not output. Default is true.
```
When we have options that are default true, then it's generally better to flip the key name to a negative, imo. So we'd go with `skip_enforcing_utf8: true`.

kaspth replied:
This was lifted from `form_for`. I figured the parity was better when it was a relatively oft unused option. Do you prefer that we rename still?

dhh replied:
This is a new API, so I saw we make the best choices we know how. Keeping parity is not important.

kaspth replied:
Sure, then we should give that treatment to these options too:

```ruby
skip_id: true # Was `include_id: false` for nested `fields`
skip_authenticity_token: true # Was just `authenticity_token: false`
```

kaspth replied:
Backing out of the `skip_authenticy_token: true` change. Since specifying a custom token via `authenticy_token: 'abcdef'` is valid. And would rather not have two options for the same thing.

dhh commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #   form_with(model: @post, url: super_posts_path)
+      #   form_with(model: @post, scope: :article)
+      #   form_with(model: @post, format: :json)
+      #   form_with(model: @post, authenticity_token: false) # Disables the token.
```
Fix when we default token to off with remote: true.

dhh commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -720,6 +961,235 @@ def fields_for(record_name, record_object = nil, options = {}, &block)
+      #       Zip code: <%= address_fields.text_field :zip_code %>
+      #     <% end %>
+      #     ...
+      #   <% end %>
```
Mixed use of ERB insertion tags. Should use them for `form_with` too.

dhh commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -720,6 +961,235 @@ def fields_for(record_name, record_object = nil, options = {}, &block)
+      #   <% end %>
+      #
+      # When address is already an association on a Person you can use
+      # +accepts_nested_attributes_for+ to define the writer method for you:
```
I'd actually like to kill `accepts_nested_attributes_for` in due time. Don't think we should promote it for this new API. Rather, let's just show how to do it by hand in the controller.

kaspth replied:
Oh yeah! 🔥 Reading through this I thought `accepts_nested_attributes_for` stuck out, refrained from doing anything because I didn't have a better idea yet. But yes, let's start down the path toward something better!

dhh replied:
For starters, I'd just drop the entire topic for now.

kaspth committed:
Link to `fields` helper method docs.

kaspth committed:
[ci skip] Documentation edits.

kaspth committed:
Use ERB tags.

kaspth commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,247 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+
+        html_options = html.merge(options.except(:index, :include_id, :builder))
+        html_options[:method] ||= :patch if model.respond_to?(:persisted?) && model.persisted?
+        html_options[:remote] = remote unless html_options.key?(:remote)
```
It's used in the `else` too as the last parameter here too: `html_options = html_options_for_form(url || {}, html_options)`

But yeah, let me see what I can do with the composition 👍

kaspth commented:
@vipulnsward @dhh thanks for the feedback, I'll look more into this tomorrow 👍

kaspth committed:
[ci skip] Remove nested attributes examples

kaspth committed:
Invert `include_id` to `skip_id`.

`skip_id: true` reads better than `include_id: false` (since the
`include_id` default is true).

kaspth committed:
Invert `remote` to `local`.

Since forms are remote by default, the option name makes more sense
as `local: true`.

kaspth committed:
Invert `enforce_utf8` to `skip_enforcing_utf8`.

kaspth committed:
Refer to the brand spanking new rails-ujs.

Soon to be bundled in Rails proper, so jquery-ujs is out.

kaspth committed:
Make `form_with` a bit more composed.

The flow is still not quite what it should be because the legacy
methods and these new ones pull at opposite ends.

Lots of options have been renamed, so now the new pieces don't fit
in so well.

I'll try to work on this in later commits after this PR (it's likely
there's a much better way to structure this whole part of Action View).

kaspth commented:
@dhh inverted the option names, removed the nested attributes example and tried to make `form_with` a bit more composed.

However, that carried over a lot of baggage from copying over `html_options_for_form` — https://github.com/rails/rails/blob/6d6249b1c1abda4f62fafcc42a7ece570c8da7e9/actionview/lib/action_view/helpers/form_tag_helper.rb#L833-L854 — which was the most sensible way to do it that I could see.

I think it would be easier to improve the code structure further here in more focused commits after this PR.

dhh commented:
Looks pretty good to me 👍

claudiob commented:
@kaspth Thanks for this PR. 

You added a `fields` method to Action View which lets you group multiple fields together, so that:

```rhtml
<%= fields :comment do |fields| %>
  <%= fields.text_field :subject %>
  <%= fields.text_field :body %>
<% end %>
```

becomes:

```
<input type="text" name="comment[subject] id="comment_subject">
<input type="text" name="comment[body] id="comment_body">
```

As you see, the HTML has lost the nuance that those two fields are _grouped together_.

In my mind, since you are grouping them together in your code, a better output would be:

```
<fieldset>
  <input type="text" name="comment[subject] id="comment_subject">
  <input type="text" name="comment[body] id="comment_body">
</fieldset>
```

since [the fieldset HTML tag](https://www.w3.org/WAI/tutorials/forms/grouping/)  is used to group related elements in a form:

> Grouping related form controls makes forms more understandable for all users, as related controls are easier to identify. It also makes it easier for people to focus on smaller and more manageable groups rather than try to grasp the entire form at once.
>
> Grouping needs to be carried out visually and in the code, for example, by using the `<fieldset>` and `<legend>` elements to associate related form controls.

@kaspth (and @dhh -- original requester for this PR)… what is your opinion on this?

Could the `fields` method introduced here have at least an option to wrap the fields in a `<fieldset>`?

kaspth commented:
I'd rather just defer any extra markup onto people themselves, so they have the complete freedom on how to design their forms.

aditya-vector commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -474,6 +474,242 @@ def apply_form_for_options!(record, object, options) #:nodoc:
+      #   simulate the verb over post.
+      # * <tt>:format</tt> - The format of the route the form submits to.
+      #   Useful when submitting to another resource type, like <tt>:json</tt>.
+      #   Skipped if a <tt>:url</tt> is passed.
```
This means when passing `:url` option, I would have to explicitly attach the format to the url as well? Wouldnt it be nicer to allow url without an explicit format and allow `:format` option too?

aledalgrande commented:
This doesn't seem to be in the edge guides, is it?
